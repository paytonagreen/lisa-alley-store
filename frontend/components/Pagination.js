import React from "react";
import PaginationStyles from "./styles/PaginationStyles";
import { useQuery, gql } from "@apollo/client";
import { perPage } from "../config";
import Head from "next/head";
import Link from "next/link";

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = ({ page }) => {
  const { data, loading, error } = useQuery(PAGINATION_QUERY);
  if (loading) return <p>Loading...</p>;
  const count = data.itemsConnection.aggregate.count;
  const pages = Math.ceil(count / perPage);
  return (
    <PaginationStyles data-test="pagination">
      <Head>
        <title>
          Lisa Alley - Page {page} of {pages}
        </title>
      </Head>
      <Link
        href={{
          pathname: "items",
          query: { page: page - 1 },
        }}
      >
        <a className="prev" aria-disabled={page <= 1}>
          Prev
        </a>
      </Link>
      <p>
        Page {page} of <span className="totalPages">{pages}</span>
      </p>
      <p>{count} Items Total</p>
      <Link
        href={{
          pathname: "items",
          query: { page: page + 1 },
        }}
      >
        <a className="next" aria-disabled={page === pages}>
          Next
        </a>
      </Link>
    </PaginationStyles>
  );
};

export default Pagination;
export { PAGINATION_QUERY };
