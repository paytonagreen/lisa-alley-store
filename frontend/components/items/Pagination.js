import React from 'react';
import PaginationStyles from '../styles/PaginationStyles';
import { useQuery, gql } from '@apollo/client';
import { perPage } from '../../config';
import Head from 'next/head';
import Link from 'next/link';
import Loader from '../utils/Loader';

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = ({ page }) => {
  const { data, loading, error } = useQuery(PAGINATION_QUERY);
  if (loading) return <Loader />;
  let count = 0;
  let pages = 0;
  if (!data && !error) {
    count = data.itemsConnection.aggregate.count;
    pages = Math.ceil(count / perPage);
    return count, pages;
  }
  ;
  return (
    <PaginationStyles data-testid="pagination">
      <Head>
        <title>
          Lisa Alley - Page {page} of {pages}
        </title>
      </Head>
      <Link
        href={{
          pathname: 'items',
          query: { page: page - 1 },
        }}
      >
        <a className="prev" aria-disabled={page <= 1}>
          Prev
        </a>
      </Link>
      <p>
        Page {page} of <span className="totalPages">{pages}</span>
      </p>
      <p>{count} Items Total</p>
      <Link
        href={{
          pathname: 'items',
          query: { page: page + 1 },
        }}
      >
        <a className="next" aria-disabled={page === pages}>
          Next
        </a>
      </Link>
    </PaginationStyles>
  );
};

export default Pagination;
export { PAGINATION_QUERY };
