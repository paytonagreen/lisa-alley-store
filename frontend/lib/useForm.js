import { useState, useEffect } from 'react';

const useForm = (callback, initValues = {}) => {
  const [values, setValues] = useState(initValues)
  const [lowercaseValues, setLowercaseValues] = useState(initValues)
  const [isSubmitting, setIsSubmitting] = useState(false);
  
  useEffect(() => {
    if(isSubmitting) {
      callback();
    }
  }, [callback, isSubmitting]);

  const handleSubmit = e => {
    if (e) {
      e.preventDefault()
    };
    setIsSubmitting(true)
  };

  const handleChange = e => {
    e.persist();
    let val = e.target.type === 'number' ? parseFloat(e.target.value) : e.target.value;
    setValues(v => ({
      ...v, [e.target.name]: val
    }))
    setLowercaseValues(v => ({
      ...v, [e.target.name]: typeof(val) === 'string' ? val.toLowerCase() : val
    }))
    
  };

  return {
    handleChange,
    handleSubmit,
    values,
    lowercaseValues
  }
}

export default useForm;