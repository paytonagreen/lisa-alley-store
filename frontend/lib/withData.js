import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/cart/Cart'
import { LOCAL_BURGER_QUERY } from '../components/burger-menu/HamburgerMenu'

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    //local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            //Read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            const data = {
              data: {cartOpen: !cartOpen},
            }
            cache.writeData(data);
            return data;
          },
          toggleBurger(_, variables, { cache }) {
            //Read the cartOpen value from the cache
            const { burgerOpen } = cache.readQuery({
              query: LOCAL_BURGER_QUERY,
            });
            const data = {
              data: {burgerOpen: !burgerOpen},
            }
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
        burgerOpen: false,
      }
    }
  });
}

export default withApollo(createClient);
