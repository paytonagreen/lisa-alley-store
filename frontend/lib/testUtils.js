import casual from 'casual';
import React from 'react';
import { render } from '@testing-library/react';
import { ApolloProvider } from '@apollo/client';
import withData from '../lib/withData';

//setup wrapper for tests//
const Providers = ({ children, apollo }) => {
  return <ApolloProvider client={apollo} addTypename={false}>{children}</ApolloProvider>;
};

const customProviders = withData(Providers);

const customRender = (ui, options) =>
  render(ui, { wrapper: customProviders, ...options });

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
  createdAt: 1,
});

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  address1: casual.address1,
  address2: '',
  city: casual.city,
  state: casual.state,
  zip: 78909,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
});

const fakeRegularUser = () => ({
  __typename: 'User',
  id: '420',
  name: casual.name,
  email: casual.email,
  address1: casual.address,
  address2: '',

  city: casual.city,
  state: casual.state,
  zip: 11111,
  permissions: ['USER'],
  orders: [],
  cart: [],
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

const fakeCartUser = () => ({
  ...fakeUser(),
  cart: [fakeCartItem()],
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeRegularUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  fakeCartUser,
  customRender as render,
};
