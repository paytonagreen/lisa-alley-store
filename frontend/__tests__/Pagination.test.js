import { screen } from '@testing-library/react';
import { render } from '../lib/testUtils';
import Pagination from '../components/items/Pagination';

const setScreenPage = (page = 1) => {
  render(
      <Pagination page={page} />
  );
};

describe('<Pagination/>', () => {
  it('displays a loading message', async () => {
    setScreenPage(1);

    expect(await screen.findByTestId('loader')).toBeInTheDocument();
  });

  it('displays pagination for 20 items', async () => {
    setScreenPage();

    expect(await screen.findByTestId('pages-index')).toHaveTextContent('Page 1 of 3');
    expect(await screen.findByTestId('pagination')).toMatchSnapshot()
    ;
  });

  it('disables prev button on first page', async () => {
    setScreenPage();

    expect(await screen.findByRole('link', {name: /Prev/i })).toHaveAttribute('aria-disabled');
  } );

  it('disables next button on last page', async () => {
    setScreenPage(4);

    expect(await screen.findByRole('link', { name: /Next/i })).toHaveAttribute('aria-disabled');
  }); 

  it('enables all buttons on a middle page', async () => {
    setScreenPage(3);

    expect(await screen.findByRole('link', { name: /Prev/i })).toHaveAttribute('aria-disabled');
    expect(await screen.findByRole('link', { name: /Next/i })).toHaveAttribute('aria-disabled');
  });
});  
