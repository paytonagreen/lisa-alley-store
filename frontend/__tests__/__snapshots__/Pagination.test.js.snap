// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> displays pagination for 20 items 1`] = `
<div
  class="PaginationStyles-sc-5gxcs1-0 kDspZA"
  data-testid="pagination"
>
  <a
    aria-disabled="true"
    class="prev"
    href="/items?page=0"
  >
    Prev
  </a>
  <p
    data-testid="pages-index"
  >
    Page 
    1
     of 
    <span
      class="totalPages"
    >
      3
    </span>
  </p>
  <p>
    20
     Items Total
  </p>
  <a
    aria-disabled="false"
    class="next"
    href="/items?page=2"
  >
    Next
  </a>
</div>
`;
